Example Best-Practices of Command-Based usage

tom131313 opened the pull request:
Add an Example Best-Practices of Command-Based usage.
Issue #6616
Reference: [CD thread](https://www.chiefdelphi.com/t/command-based-best-practices-for-2025-community-feedback/465602)
Utilize an LED-based system to show various best-practices in coding Command-Based programs.

illinar commented on wpilibjExamples/src/main/java/edu/wpi/first/wpilibj/examples/commandbasedbestpracticeled/subsystems/RobotSignalsLEDbufferLEDSubsystem.java:
```diff
@@ -0,0 +1,29 @@
+    buffer = new AddressableLEDBuffer(length);
+
+    // start the physical LED strip output
+    periodicTask.register(()->strip.start(), 0.1, 0.01);
+  }
+
+  public void periodic() {
+    strip.setData(buffer);
+  }
```
Did you mean to put `strip.setData(buffer)` in the periodic callback and remove `periodic()` altogether? `strip.start()` should be invoked once. Any specific reason to offset by 1ms and run at 100ms?

illinar commented on wpilibjExamples/src/main/java/edu/wpi/first/wpilibj/examples/commandbasedbestpracticeled/subsystems/RobotSignalsLEDbufferLEDSubsystem.java:
```diff
@@ -0,0 +1,29 @@
+import edu.wpi.first.wpilibj.AddressableLEDBuffer;
+import frc.robot.PeriodicTask;
+
+public class RobotSignalsLEDbufferLEDSubsystem  {
```
Not sure why such a verbose name, also I would not include `Subsystem` suffix since this is no longer a subsystem. How about `LEDResource`?

tom131313 replied:
I apologize for leaving some trash from the start and then got distracted trying to get output from the example and getting it into a WPILib Pull Request.

My refactored version is available [here](https://github.com/tom131313/allwpilib/tree/main/wpilibjExamples/src/main/java/edu/wpi/first/wpilibj/examples/commandbasedbestpracticeled). Believe it or not it does resemble the original except I have added working output with commentary and didn't pursue the anticipated new AddressableLED functions which we can add when they are readily available. (Xbox controller buttons `A`, `B`, and `X` do something in different robot modes. Now I see that I should document the expected behavior.)

I can't remember - are we including the word `subsystem` in the Java class/file name or not? I left in `subsystem` but scrubbed `Command` and `Trigger` from names.)

I would appreciate your and @Oblarg comments on what can be done better as a best practice. Thanks.

tom131313 replied:
@illinar I have completed a working example LED resource/subsystem/view based about 99% on your example code.

I did find @SamCarlberg nice addition to the LED process that you recommended. I copied his entire work into my example since I don't see any other way to reference it in my code. All those files would have to be deleted from my example and appropriate `imports` made to the WPILib when it has those files available.

My changes to your code are trivial to make it a little easier to test. I simulated a couple of events with Xbox controller button presses and changed an `onTrue` to `whileTrue`. I added a third view of the LEDs only to display your dynamic Enable/Disable Patterns.

Since the basis of all LED addressable views is the same I have only one class for the LEDview and instantiate it three times for the Top, Main, and EnableDisable views.

I welcome comments from all - you, @Oblarg and anyone to further refine structure.

Thanks for your contribution of a fine example.

illinar commented on wpilibjExamples/src/main/java/edu/wpi/first/wpilibj/examples/commandbasedbestpracticeled/PeriodicTask.java:
```diff
@@ -0,0 +1,24 @@
+
+ */
+@FunctionalInterface
+public interface PeriodicTask {
```
I feel like this should be part of WPILIB API surface if this pattern is an officially blessed mechanism of injecting periodic processing support.

calcmogul replied:
> if this pattern is an officially blessed mechanism of injecting periodic processing support

It's not. We have TimedRobot.addPeriodic() instead.

tom131313 commented:
@jasondaming I happened to see that my PR wants WPILib style so I've changed the code to satisfy at least most of the messages that I found. I see a `wpiformat` log that said at the end it was in error but I don't know how to find the errors in 3000 lines of output (tried to look but failed).

I don't know how to trigger the style checking again for round two of changing the style. Was that triggered by you? Thanks

calcmogul commented:
To fix C++:
```
pip install wpiformat
cd allwpilib
python -m wpiformat
```
To fix Java, remove the commented out import line in `wpilibjExamples/src/main/java/edu/wpi/first/wpilibj/examples/commandbasedbestpracticeled/LEDPattern.java`, then run `./gradlew wpilibjExamples:spotlessApply`.

I'd fix it for you, but you set your PR to not allow pushes by maintainers.

tom131313 commented:
@calcmogul
> run `./gradlew wpilibjExamples:spotlessApply`

I corrected (most) problems with the formatting from your running `spotless`. I also fixed the `CR/LF` issue but not sure if I did it in the right place.

I ran `spotless` as you said and can't see any issues and that cannot possibly be right. I looked at everything at the link to the `gradle build scan`.

Several days ago with lots of help I got the `allwpilib fork` for the PR to my `github` account. If there is something I need to change to give update access to all who need it, I'd be happy to follow more instructions.

More help is appreciated. Thanks.

My terminal output:
> Microsoft Windows [Version 10.0.19045.4412]
> (c) Microsoft Corporation. All rights reserved.
> 
> C:\Users\RKT\frc\FRC2024\Code\allwpilib>gradlew wpilibjExamples:spotlessApply
> 
> > Configure project :
> Tag is null. This does not match the expected version number pattern.
> No version number was generated.
> Skipping builds for arm32 (toolchain is marked optional)
> Skipping builds for arm64 (toolchain is marked optional)
> 
> BUILD SUCCESSFUL in 24s
> 13 actionable tasks: 2 executed, 11 up-to-date
> 
> Publishing build scan...
> https://gradle.com/s/kpdv2euzlyeqc

WispySparks commented:
Why are all of the LED classes copied? (AddressableLED, AddressableLEDBuffer, AddressableLEDBufferView, Color, Color8Bit, LEDPattern, LEDReader, and LEDWriter)

tom131313 commented:
> Why are all of the LED classes copied? (AddressableLED, AddressableLEDBuffer, AddressableLEDBufferView, Color, Color8Bit, LEDPattern, LEDReader, and LEDWriter)

Illinar suggested using them and I agreed to that. They aren't part of the stable WPILib distribution that I have so I copied them from their PR. They would be removed from this example directory when they are available to import from their final resting place.

I would appreciate someone running `spotless` again to see if I caught all of its requirements. It appears the instructions I got are for Linux which I do not have. I installed `spotless` on my VSCode Windows PC and it gave different messages than the one from WPILib usage. It wanted all of the `m_` name prefixes removed.

WispySparks commented:
> Illinar suggested using them and I agreed to that. They aren't part of the stable WPILib distribution that I have so I copied them from their PR. They would be removed from this example directory when they are available to import from their final resting place.
They've since been merged, you should merge main and delete the local versions.

tom131313 commented:
I've taken this program with a few examples of using command-based as far as I can - it accomplishes all I intended. I can't go any further without WPILib reviewer input. If these examples are of value to WPILib, I'd be happy to recode whatever needs to make it a real WPILib example. The name of the folder should be changed at least to remove the "word" led from it as that is misleading and off-putting. I used LEDs in place of the PWM motor controllers or other devices merely to show actions of the examples.

If these examples aren't what is wanted, I'll close the PR and delete the repository from my account.

Gold856 reviewed (changes requested):
The overall structure of this example confuses me a lot. A big overarching theme is that subsystems are used by other subsystems. This feels contradictory to the entire point of the Command-based framework. Why are LEDViews, which are also subsystems, created in RobotSignals, then passed to each subsystem? Why does GroupDisjointTest contain multiple instances of GroupDisjoint? What is GroupDisjointTest trying to show me? There are so many commands, and I'm not sure what they're doing or what it's supposed to demonstrate.

The MooreLikeFSM subsystem also confuses me. It's contradictory to HistoryFSM where it says that commands should not schedule other commands, yet that's what MooreLikeFSM does. MooreLikeFSM also seems like it would be better represented as a giant sequential command since it seems the entire structure is just running commands one after another.

spacey-sooty reviewed (changes requested):
I agree with all of @Gold856 comments but I also don't see the benefit in adding a new example to show command based best practices. The RapidReactBot example has been used as the up to date and best practice example up until this point in time and I don't see why we should change that. If theres anything that that does incorrectly that should be updated instead of creating a new example. We already have issues keeping current examples up to date, why make that problem worse?

tom131313 commented:
These examples of Command-Based programming are intended to illustrate acceptable practices that tend to push usage to all the edges of somewhat unusual usage. Thus, the comments are all valid in that there are multiple examples to sort through and they don't seem to illustrate the normal best practice. Since confusion is anticipated and that is counter to the intention, this project will be dropped and closed.
